File: app/src/test/java/kittentrate/data/domain/GameManagerTest.java
package kittentrate.data.domain;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import kittentrate.game.Card;
import kittentrate.game.GameManager;
import kittentrate.game.GamePresenter;

import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by Manuel Lorenzo
 */
public class GameManagerTest {
    @Mock
    private GamePresenter gamePresenter;
    @Mock
    private Card card1;
    @Mock
    private Card card2;

    private GameManager gameManager;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        gameManager = new GameManager(gamePresenter);
    }

    @Test
    public void testGameManager_gamePresenterShouldTurnCardsOver_whenCardsAreFlipped() throws Exception {
        when(card1.getId()).thenReturn("1");
        when(card2.getId()).thenReturn("2");
        gameManager.cardFlipped(0, card1);
        gameManager.cardFlipped(1, card2);

        verify(gamePresenter).onTurnCardsOver();
    }

    @Test
    public void testGameManager_gamePresenterShouldRemoveViewFlipperAndChangeScore_whenCardsAreMatched() throws Exception {
        when(card1.getId()).thenReturn("1");

        gameManager.cardFlipped(0, card1);
        gameManager.cardFlipped(1, card1);
        verify(gamePresenter, times(2)).removeViewFlipper();
        verify(gamePresenter, times(2)).notifyAdapterItemRemoved(anyString());
        verify(gamePresenter).onGameScoreChanged(anyInt());
    }

    @Test
    public void testGameManager_gamePresenterShouldNotifyItemsChanged_whenGameManagerFlipsCard() throws Exception {
        when(card1.getId()).thenReturn("1");
        gameManager.cardFlipped(0, card1);
        gameManager.removeCardsFromMap();

        verify(gamePresenter).notifyAdapterItemChanged(anyInt());
    }
}

File: app/src/test/java/kittentrate/data/domain/GameManagerTest.kt
package kittentrate.data.domain

import kittentrate.RobolectricTestHelper
import kittentrate.ui.game.Card
import kittentrate.ui.game.GameManager
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations
import org.robolectric.RobolectricTestRunner

/**
 * Created by Manuel Lorenzo
 */
@RunWith(RobolectricTestRunner::class)
class GameManagerTest : RobolectricTestHelper() {
    @Mock
    private val gamePresenter: GamePresenter? = null
    @Mock
    private val card1: Card? = null
    @Mock
    private val card2: Card? = null

    private var gameManager: GameManager? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        gameManager = GameManager(gamePresenter!!)
    }

    @Test
    @Throws(Exception::class)
    fun testGameManager_gamePresenterShouldTurnCardsOver_whenCardsAreFlipped() {
        `when`(card1!!.id).thenReturn("1")
        `when`(card2!!.id).thenReturn("2")
        gameManager!!.cardFlipped(0, card1)
        gameManager!!.cardFlipped(1, card2)

        verify<GamePresenter>(gamePresenter).onTurnCardsOver()
    }

    @Test
    @Throws(Exception::class)
    fun testGameManager_gamePresenterShouldRemoveViewFlipperAndChangeScore_whenCardsAreMatched() {
        `when`(card1!!.id).thenReturn("1")

        gameManager!!.cardFlipped(0, card1)
        gameManager!!.cardFlipped(1, card1)
        verify<GamePresenter>(gamePresenter, times(2)).removeViewFlipper()
        verify<GamePresenter>(gamePresenter, times(2)).notifyAdapterItemRemoved(anyString())
        verify<GamePresenter>(gamePresenter).onGameScoreChanged(anyInt())
    }

    @Test
    @Throws(Exception::class)
    fun testGameManager_gamePresenterShouldNotifyItemsChanged_whenGameManagerFlipsCard() {
        `when`(card1!!.id).thenReturn("1")
        gameManager!!.cardFlipped(0, card1)
        gameManager!!.removeCardsFromMap()

        verify<GamePresenter>(gamePresenter).notifyAdapterItemChanged(anyInt())
    }
}